cpp/src/copying/contiguous_split.cu

```diff
@@ -141,28 +279,164 @@ struct column_copy_functor {
+   out_cols.push_back(column_view(in.type(), in.size(), nullptr,
+                                    validity_buf, UNKNOWN_NULL_COUNT, 0,
+                                    { out_offsets, out_chars }));
+}
+
+/**
+ * @brief Information about a string column in a table view.
+ * 
+ * Used internally by preprocess_string_column_info as part of a device-accessible 
+ * vector for computing final string information in a single kernel call.
+ */
+struct column_preprocess_info {
+   size_type                  index;
+   size_type                  offset;
+   size_type                  size;
+   bool                       nullable;      
+   cudf::column_device_view   offsets;
+};
+
+/**
+ * @brief Functor called by the `type_dispatcher` to collect string columns.
+ */
+struct string_column_collector {
+   template <typename T>
+   void operator()(cudf::column_view const& c, thrust::host_vector<column_preprocess_info>& offset_columns, size_type column_index){}   
+};
+template<>
+void string_column_collector::operator()<string_view>(cudf::column_view const& c, thrust::host_vector<column_preprocess_info>& offset_columns, size_type column_index)
+{
+   // constructing device view from the offsets column only, because doing so for the entire
+   // strings_column_view will result in memory allocation/cudaMemcpy() calls, which would
+   // defeat the whole purpose of this step.
+   cudf::column_device_view cdv((strings_column_view(c)).offsets(), 0, 0);
+   offset_columns.push_back(column_preprocess_info{column_index, c.offset(), c.size(), c.nullable(), cdv});      
+}
+
+/**
+ * @brief Preprocess information about all strings columns in a table view.
+ * 
+ * In order to minimize how often we touch the gpu, we need to preprocess various pieces of information
+ * about the string columns in a table as a batch process.  This function builds a list of the offset
+ * columns for all input string columns and computes this information with a single thrust call.  In addition,
+ * the vector returned is allocated for -all- columns in the table so further processing of non-string columns
+ * can happen afterwards.
+ * 
+ * The key things this function avoids
+ * - avoiding reaching into gpu memory on the cpu to retrieve offsets to compute string sizes.
+ * - creating column_device_views on the base string_column_view itself as that causes gpu memory allocation.
+ */
+thrust::host_vector<column_split_info> preprocess_string_column_info(cudf::table_view const& t, thrust::device_vector<column_split_info>& device_split_info, cudaStream_t stream)
+{          
+   // build a list of all the offset columns and their indices for all input string columns and put them on the gpu   
+   thrust::host_vector<column_preprocess_info> offset_columns;
+   offset_columns.reserve(t.num_columns());  // worst case
+   size_type column_index = 0;
+
+   // collect only string columns
+   std::for_each(t.begin(), t.end(), [&offset_columns, &column_index](cudf::column_view const& c){
```

Comment:
As discussed offline, there's some weird issue here with using std::back_inserter() as well as issues preallocating a max-sized array and just using offset_columns.begin().

One option here is to revert back to what I was using earlier, which is just a simple for_each() that checks for STRING directly instead of going through the type dispatcher.

```
  std::for_each(t.begin(), t.end(), [&offset_columns, &column_index](cudf::column_view const& c){
      if(c.type().id() == STRING){         
         cudf::column_device_view cdv((strings_column_view(c)).offsets(), 0, 0);
         offset_columns.push_back(column_preprocess_info{column_index, c.offset(), c.size(), c.nullable(), cdv});
      }
      column_index++;
   });   
```

It's not as nice as transform / transform_if but it's pretty simple.  Thoughts?

Reply:
Sure, that's a fair compromise.